## Model Profiling with 5 iterations of Warmup
| size   | mode    | device   |   batch_size |   context_length |   d_model |   d_ff |   num_layers |   num_heads |   warmup_iter |   n_iter |    mean_s |       std_s |   peak_mem_mb |
|:-------|:--------|:---------|-------------:|-----------------:|----------:|-------:|-------------:|------------:|--------------:|---------:|----------:|------------:|--------------:|
| small  | fwd     | cuda     |            4 |              128 |       768 |   3072 |           12 |          12 |             5 |       10 | 0.022593  | 0.000104068 |       550.108 |
| small  | fwd+bwd | cuda     |            4 |              128 |       768 |   3072 |           12 |          12 |             5 |       10 | 0.0709633 | 0.00825605  |      1280.44  |
| medium | fwd     | cuda     |            4 |              128 |      1024 |   4096 |           24 |          16 |             5 |       10 | 0.0445866 | 0.000133085 |      1712.23  |
| medium | fwd+bwd | cuda     |            4 |              128 |      1024 |   4096 |           24 |          16 |             5 |       10 | 0.115326  | 0.0012455   |      3422.45  |
| large  | fwd     | cuda     |            4 |              128 |      1280 |   5120 |           36 |          20 |             5 |       10 | 0.0805293 | 0.00250605  |      3910.05  |
| large  | fwd+bwd | cuda     |            4 |              128 |      1280 |   5120 |           36 |          20 |             5 |       10 | 0.143583  | 0.0150844   |      7743.89  |
| xl     | fwd     | cuda     |            4 |              128 |      1600 |   6400 |           48 |          25 |             5 |       10 | 0.066871  | 2.56107e-05 |      7918.65  |
| xl     | fwd+bwd | cuda     |            4 |              128 |      1600 |   6400 |           48 |          25 |             5 |       10 | 0.214561  | 0.000210963 |     15732.7   |
| 2.7B   | fwd     | cuda     |            4 |              128 |      2560 |  10240 |           32 |          32 |             5 |       10 | 0.0878882 | 0.000122993 |     13268.7   |
| 2.7B   | fwd+bwd | cuda     |            4 |              128 |      2560 |  10240 |           32 |          32 |             5 |       10 | 0.303348  | 0.000622047 |     26302.6   |

## No Warmup
| size   | mode    | device   |   batch_size |   context_length |   d_model |   d_ff |   num_layers |   num_heads |   warmup_iter |   n_iter |    mean_s |      std_s |   peak_mem_mb |
|:-------|:--------|:---------|-------------:|-----------------:|----------:|-------:|-------------:|------------:|--------------:|---------:|----------:|-----------:|--------------:|
| small  | fwd     | cuda     |            4 |              128 |       768 |   3072 |           12 |          12 |             0 |       10 | 0.11472   | 0.266369   |       550.108 |
| small  | fwd+bwd | cuda     |            4 |              128 |       768 |   3072 |           12 |          12 |             0 |       10 | 0.0773883 | 0.048441   |      1280.44  |
| medium | fwd     | cuda     |            4 |              128 |      1024 |   4096 |           24 |          16 |             0 |       10 | 0.04986   | 0.0169794  |      1712.23  |
| medium | fwd+bwd | cuda     |            4 |              128 |      1024 |   4096 |           24 |          16 |             0 |       10 | 0.107343  | 0.0113035  |      3422.45  |
| large  | fwd     | cuda     |            4 |              128 |      1280 |   5120 |           36 |          20 |             0 |       10 | 0.0407184 | 0.00957906 |      3910.05  |
| large  | fwd+bwd | cuda     |            4 |              128 |      1280 |   5120 |           36 |          20 |             0 |       10 | 0.146847  | 0.0254056  |      7743.89  |
| xl     | fwd     | cuda     |            4 |              128 |      1600 |   6400 |           48 |          25 |             0 |       10 | 0.0701499 | 0.0105331  |      7918.65  |
| xl     | fwd+bwd | cuda     |            4 |              128 |      1600 |   6400 |           48 |          25 |             0 |       10 | 0.220418  | 0.021678   |     15732.7   |
| 2.7B   | fwd     | cuda     |            4 |              128 |      2560 |  10240 |           32 |          32 |             0 |       10 | 0.0928832 | 0.0158411  |     13268.7   |
| 2.7B   | fwd+bwd | cuda     |            4 |              128 |      2560 |  10240 |           32 |          32 |             0 |       10 | 0.302375  | 0.00282029 |     26303.3   |

## 1 iteration Warmup
| size   | mode    | device   |   batch_size |   context_length |   d_model |   d_ff |   num_layers |   num_heads |   warmup_iter |   n_iter |    mean_s |       std_s |   peak_mem_mb |
|:-------|:--------|:---------|-------------:|-----------------:|----------:|-------:|-------------:|------------:|--------------:|---------:|----------:|------------:|--------------:|
| small  | fwd     | cuda     |            4 |              128 |       768 |   3072 |           12 |          12 |             1 |       10 | 0.0263708 | 0.000674041 |       550.108 |
| small  | fwd+bwd | cuda     |            4 |              128 |       768 |   3072 |           12 |          12 |             1 |       10 | 0.0700107 | 0.00302113  |      1280.44  |
| medium | fwd     | cuda     |            4 |              128 |      1024 |   4096 |           24 |          16 |             1 |       10 | 0.0498092 | 0.000133697 |      1712.23  |
| medium | fwd+bwd | cuda     |            4 |              128 |      1024 |   4096 |           24 |          16 |             1 |       10 | 0.107668  | 0.0102769   |      3422.45  |
| large  | fwd     | cuda     |            4 |              128 |      1280 |   5120 |           36 |          20 |             1 |       10 | 0.0370991 | 0.000139602 |      3910.05  |
| large  | fwd+bwd | cuda     |            4 |              128 |      1280 |   5120 |           36 |          20 |             1 |       10 | 0.140533  | 0.0149137   |      7743.89  |
| xl     | fwd     | cuda     |            4 |              128 |      1600 |   6400 |           48 |          25 |             1 |       10 | 0.066688  | 0.000166981 |      7918.65  |
| xl     | fwd+bwd | cuda     |            4 |              128 |      1600 |   6400 |           48 |          25 |             1 |       10 | 0.214248  | 0.000345418 |     15732.7   |
| 2.7B   | fwd     | cuda     |            4 |              128 |      2560 |  10240 |           32 |          32 |             1 |       10 | 0.0873711 | 0.000188711 |     13268.7   |
| 2.7B   | fwd+bwd | cuda     |            4 |              128 |      2560 |  10240 |           32 |          32 |             1 |       10 | 0.301108  | 4.89181e-05 |     26302.6   |

## 2 iteration Warmup Step
| size   | mode    | device   |   batch_size |   context_length |   d_model |   d_ff |   num_layers |   num_heads |   warmup_iter |   n_iter |    mean_s |       std_s |   peak_mem_mb |
|:-------|:--------|:---------|-------------:|-----------------:|----------:|-------:|-------------:|------------:|--------------:|---------:|----------:|------------:|--------------:|
| small  | fwd     | cuda     |            4 |              128 |       768 |   3072 |           12 |          12 |             2 |       10 | 0.0127079 | 0.00020617  |       550.108 |
| small  | fwd+bwd | cuda     |            4 |              128 |       768 |   3072 |           12 |          12 |             2 |       10 | 0.0678351 | 0.00920366  |      1280.44  |
| medium | fwd     | cuda     |            4 |              128 |      1024 |   4096 |           24 |          16 |             2 |       10 | 0.0250819 | 0.000291259 |      1712.23  |
| medium | fwd+bwd | cuda     |            4 |              128 |      1024 |   4096 |           24 |          16 |             2 |       10 | 0.108795  | 0.00971464  |      3422.45  |
| large  | fwd     | cuda     |            4 |              128 |      1280 |   5120 |           36 |          20 |             2 |       10 | 0.0376591 | 0.000149408 |      3910.05  |
| large  | fwd+bwd | cuda     |            4 |              128 |      1280 |   5120 |           36 |          20 |             2 |       10 | 0.140864  | 0.0137292   |      7743.89  |
| xl     | fwd     | cuda     |            4 |              128 |      1600 |   6400 |           48 |          25 |             2 |       10 | 0.0670614 | 0.000129    |      7918.65  |
| xl     | fwd+bwd | cuda     |            4 |              128 |      1600 |   6400 |           48 |          25 |             2 |       10 | 0.215266  | 0.000679135 |     15732.7   |
| 2.7B   | fwd     | cuda     |            4 |              128 |      2560 |  10240 |           32 |          32 |             2 |       10 | 0.0875296 | 0.000115729 |     13268.7   |
| 2.7B   | fwd+bwd | cuda     |            4 |              128 |      2560 |  10240 |           32 |          32 |             2 |       10 | 0.302965  | 0.00062237  |     26302.6   |


### Why warmup matters:

Eliminates one-time initialization costs (CUDA kernel compilation, memory allocation, GPU frequency ramp-up)
Reduces variance from 200%+ down to <1% of mean execution time
Particularly critical for smaller models where overhead is proportionally larger

### Warmup iteration comparison:

0 iterations: Extremely high variance (up to 231% relative std)
1-2 iterations: Major improvement, ~1-3% variance - acceptable for most cases
5 iterations: Optimal stability, <1% variance - best for rigorous benchmarking

### Key insight:

GPUs need time to reach stable clock frequencies and for CUDA runtime to optimize
1-2 warmups handle most initialization, but 5 iterations ensure complete stabilization
The small cost of extra warmup iterations pays off with significantly more reliable measurements

### bfloat16 autocast with warmup
| size   | mode    | device   |   batch_size |   context_length |   d_model |   d_ff |   num_layers |   num_heads |   warmup_iter |   n_iter |    mean_s |       std_s |   peak_mem_mb |
|:-------|:--------|:---------|-------------:|-----------------:|----------:|-------:|-------------:|------------:|--------------:|---------:|----------:|------------:|--------------:|
| small  | fwd     | cuda     |            4 |              128 |       768 |   3072 |           12 |          12 |             5 |       10 | 0.0139131 | 0.000193304 |       797.132 |
| small  | fwd+bwd | cuda     |            4 |              128 |       768 |   3072 |           12 |          12 |             5 |       10 | 0.0791648 | 0.00339954  |      1262.96  |
| medium | fwd     | cuda     |            4 |              128 |      1024 |   4096 |           24 |          16 |             5 |       10 | 0.0282443 | 0.000629777 |      2480.99  |
| medium | fwd+bwd | cuda     |            4 |              128 |      1024 |   4096 |           24 |          16 |             5 |       10 | 0.123578  | 0.00524961  |      3579.45  |
| large  | fwd     | cuda     |            4 |              128 |      1280 |   5120 |           36 |          20 |             5 |       10 | 0.0422451 | 0.000129378 |      5708.23  |
| large  | fwd+bwd | cuda     |            4 |              128 |      1280 |   5120 |           36 |          20 |             5 |       10 | 0.162037  | 0.0151045   |      7784.98  |
| xl     | fwd     | cuda     |            4 |              128 |      1600 |   6400 |           48 |          25 |             5 |       10 | 0.0556125 | 0.000107493 |     11754     |
| xl     | fwd+bwd | cuda     |            4 |              128 |      1600 |   6400 |           48 |          25 |             5 |       10 | 0.216072  | 0.0245934   |     15749.9   |
| 2.7B   | fwd     | cuda     |            4 |              128 |      2560 |  10240 |           32 |          32 |             5 |       10 | 0.0383748 | 0.000362194 |     19655     |
| 2.7B   | fwd+bwd | cuda     |            4 |              128 |      2560 |  10240 |           32 |          32 |             5 |       10 | 0.149899  | 0.0150371   |     26292.7   |

At lower model sizes, overhead is larger than the actual execution, so bf16 does not provide much benefit. But at 2.7B, there is almost 50% improvement in speed, and the gap probably would increase dramatically on even larger models (since FLOPs is a magnitude higher with bf16 than fp32)

In [1]:
import torch

In [3]:
s = torch.tensor(0,dtype=torch.float32)
for i in range(1000):
    s += torch.tensor(0.01,dtype=torch.float32)
print(s)
s = torch.tensor(0,dtype=torch.float16)
for i in range(1000):
    s += torch.tensor(0.01,dtype=torch.float16)
print(s)
s = torch.tensor(0,dtype=torch.float32)
for i in range(1000):
    s += torch.tensor(0.01,dtype=torch.float16)
print(s)
s = torch.tensor(0,dtype=torch.float32)
for i in range(1000):
    x = torch.tensor(0.01,dtype=torch.float16)
    s += x.type(torch.float32)
print(s)

tensor(10.0001)
tensor(9.9531, dtype=torch.float16)
tensor(10.0021)
tensor(10.0021)


Case 1: Pure float32
All operations are in float32 (32-bit precision)
Should theoretically be 10.0, but accumulated rounding errors from 1000 additions result in 10.0001
Float32 has ~7 decimal digits of precision

Case 2: Pure float16
All operations are in float16 (16-bit precision, also called "half precision")
Float16 has only ~3-4 decimal digits of precision
Much larger accumulated error: 9.9531 instead of 10.0
The value 0.01 cannot be represented exactly in float16, and errors compound quickly

Case 3: Mixed Type with automatic promotion
Key behavior: PyTorch promotes float16 to float32 during the addition
However, 0.01 is first converted to float16 (with precision loss), then promoted to float32
This "bad" float16 representation is added to the float32 accumulator
Result: worse than pure float32, but the accumulator maintains float32 precision

Case 4: Same as Case 3

In [12]:
import torch
import torch.nn as nn
from torch import Tensor

class ToyModel(nn.Module):
    def __init__(self, in_features: int, out_features: int):
        super().__init__()
        self.fc1 = nn.Linear(in_features, 10, bias=False)
        self.ln = nn.LayerNorm(10)
        self.fc2 = nn.Linear(10, out_features, bias=False)
        self.relu = nn.ReLU()
    def forward(self, x):
        print(f"input x: {x.dtype}")
        x = self.fc1(x)
        print(f"post fc1 x: {x.dtype}")
        x = self.relu(x)
        print(f"post relu x: {x.dtype}")
        x = self.ln(x)
        print(f"post ln x: {x.dtype}")
        x = self.fc2(x)
        print(f"post fc2 x: {x.dtype}")
        return x

x = torch.randn(10, dtype=torch.float32, device='mps')
model = ToyModel(10, 10).to('mps')
model.zero_grad(set_to_none=True)

print("Without Autocast:")
y = model(x)
loss = y.sum()
loss.backward()
print(f"grad data type {model.fc1.weight.grad.dtype}")

print("With Autocast:")
model.zero_grad(set_to_none=True)
with torch.autocast(device_type='mps', dtype=torch.float16):
    y=model(x)
    loss = y.sum()
    loss.backward()
    print(f"grad data type {model.fc1.weight.grad.dtype}")

print("With Autocast bf16 (limited support on mps):")
model.zero_grad(set_to_none=True)
with torch.autocast(device_type='mps', dtype=torch.bfloat16):
    y=model(x)
    loss = y.sum()
    loss.backward()
    print(f"grad data type {model.fc1.weight.grad.dtype}")

Without Autocast:
input x: torch.float32
post fc1 x: torch.float32
post relu x: torch.float32
post ln x: torch.float32
post fc2 x: torch.float32
grad data type torch.float32
With Autocast:
input x: torch.float32
post fc1 x: torch.float16
post relu x: torch.float16
post ln x: torch.float32
post fc2 x: torch.float16
grad data type torch.float32
With Autocast bf16:
input x: torch.float32
post fc1 x: torch.bfloat16
post relu x: torch.bfloat16
post ln x: torch.float32
post fc2 x: torch.bfloat16
grad data type torch.float32


Data Types in Mixed Precision Training

Model Parameters (within autocast context):
FP32 - The original parameters remain in FP32. Autocast doesn't change the stored parameter types; it only casts them temporarily during forward pass operations.

Output of fc1 (first linear layer):
FP16 - Linear layers are in the autocast "eligible" list, so the FP32 weights are cast to FP16, and the computation happens in FP16.

Output after ReLU:
FP16 - ReLU is a pointwise operation that preserves the input dtype, so it remains FP16.

Output of LayerNorm (ln):
FP32 - LayerNorm is on the "FP32 ops" list because it involves reductions and normalization that benefit from higher precision for numerical stability.

Output of fc2 (second linear layer):
FP16 - Even though the input is FP32 from LayerNorm, fc2 is an eligible operation, so it gets cast back to FP16 for the matrix multiplication.

Model's predicted logits:
FP16 - The final output from fc2 remains in FP16.

Loss:
FP32 - Most loss functions (CrossEntropyLoss, MSELoss, etc.) automatically promote their inputs to FP32 for numerical stability, even within the autocast context.

Model's gradients:
FP32 - Gradients are accumulated in the same dtype as the parameters, which remain in FP32. During the backward pass, even though some forward activations were FP16, the gradients are computed and accumulated in FP32.

What parts of LayerNorm are sensitive to mixed precision?

LayerNorm logic:


In [13]:
def layer_norm(x, gamma, beta, eps=1e-5):
    # Normalize over last dimension(s)
    mean = x.mean(dim=-1, keepdim=True)
    var = x.var(dim=-1, unbiased=False, keepdim=True)
    x_norm = (x - mean) / torch.sqrt(var + eps)
    
    # Apply affine transform
    return gamma * x_norm + beta

Variance computation, division by small numbers are susceptible to low precision errors. Even grads in these case would involves multiplicationa and division which could cause issues in grads.

bfloat16 has a higher range so autocast actually keeps bf16 for LayerNorm. Not true for MPS

## Vanilla Attention Benchmarks
| device   |   batch_size |   d_model |   seq_len |   fw_mean_s |    fw_std_s |   bw_mean_s |    bw_std_s |   mem_before_bwd_mb | status   | error   |
|:---------|-------------:|----------:|----------:|------------:|------------:|------------:|------------:|--------------------:|:---------|:--------|
| cuda     |            8 |        16 |       256 | 0.00257019  | 0.0216639   |  0.00340066 | 0.0286695   |             69.7744 | ok       |         |
| cuda     |            8 |        16 |      1024 | 0.000339281 | 0.000260863 |  0.0011211  | 0.000248008 |            135.095  | ok       |         |
| cuda     |            8 |        16 |      4096 | 0.0025646   | 0.000202286 |  0.0062171  | 8.71038e-05 |           1116.38   | ok       |         |
| cuda     |            8 |        16 |      8192 | 0.00994167  | 7.48688e-05 |  0.0238352  | 0.000104717 |           4216.75   | ok       |         |
| cuda     |            8 |        16 |     16384 | 0.0390492   | 0.00103785  |  0.0940123  | 0.000161966 |          16561.5    | ok       |         |
| cuda     |            8 |        16 |     32768 |            |            |            |            |                     | oom      | OOM     |
| cuda     |            8 |        32 |       256 | 0.000114554 | 9.4068e-05  |  0.00115204 | 0.00025819  |             71.5244 | ok       |         |
| cuda     |            8 |        32 |      1024 | 0.000256942 | 0.000121982 |  0.00114297 | 0.000280624 |            142.095  | ok       |         |
| cuda     |            8 |        32 |      4096 | 0.00265978  | 8.06978e-05 |  0.00630066 | 8.19027e-05 |           1144.38   | ok       |         |
| cuda     |            8 |        32 |      8192 | 0.0102459   | 6.41021e-05 |  0.0241119  | 8.92101e-05 |           4272.75   | ok       |         |
| cuda     |            8 |        32 |     16384 | 0.0400996   | 7.79182e-05 |  0.095068   | 0.000117993 |          16673.5    | ok       |         |
| cuda     |            8 |        64 |       256 | 0.000121771 | 0.000129119 |  0.00112999 | 0.000173525 |             75.0244 | ok       |         |
| cuda     |            8 |        64 |      1024 | 0.000267888 | 6.20349e-05 |  0.00111495 | 0.000106903 |            156.095  | ok       |         |
| cuda     |            8 |        64 |      4096 | 0.00292579  | 7.5545e-05  |  0.00680642 | 8.12444e-05 |           1200.38   | ok       |         |
| cuda     |            8 |        64 |      8192 | 0.011448    | 6.54827e-05 |  0.0264497  | 0.000127748 |           4384.75   | ok       |         |
| cuda     |            8 |        64 |     16384 | 0.0454684   | 7.17934e-05 |  0.104967   | 0.000115151 |          16897.5    | ok       |         |
| cuda     |            8 |        64 |     32768 |            |            |            |            |                     | oom      | OOM     |
| cuda     |            8 |       128 |       256 | 0.000127997 | 0.000138762 |  0.0011494  | 0.000274242 |             82.0244 | ok       |         |
| cuda     |            8 |       128 |      1024 | 0.000306624 | 6.59696e-05 |  0.00117051 | 0.000109251 |            184.095  | ok       |         |
| cuda     |            8 |       128 |      4096 | 0.00356625  | 0.000295742 |  0.00803529 | 9.82754e-05 |           1312.38   | ok       |         |
| cuda     |            8 |       128 |      8192 | 0.0137337   | 6.37877e-05 |  0.0308589  | 8.39673e-05 |           4608.75   | ok       |         |
| cuda     |            8 |       128 |     16384 | 0.0542224   | 8.8418e-05  |  0.122406   | 0.000125221 |          17345.5    | ok       |         |
| cuda     |            8 |       128 |     32768 |            |            |            |            |                     | oom      | OOM     |
| cuda     |            8 |       256 |     16384 | 0.0743859   | 5.71176e-05 |  0.163467   | 5.38268e-05 |           18241.5   | ok       |         |
| cuda     |            8 |       256 |     32768 |            |            |            |            |                     | oom      | OOM     |
| cuda     |            8 |       512 |     32768 |            |            |            |            |                     | oom      | OOM     |
| cuda     |            8 |      1024 |     16384 | 0.204574    | 0.000816931 |  0.414481   | 0.000385968 |          23617.5    | ok       |         |
| cuda     |            8 |      1024 |     32768 |            |            |            |            |                     | oom      | OOM     |